sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -618,19 +618,41 @@ object FunctionRegistry {
           }
           throw new AnalysisException(invalidArgumentsMsg)
         }
-        Try(f.newInstance(expressions : _*).asInstanceOf[Expression]) match {
-          case Success(e) => e
-          case Failure(e) =>
-            // the exception is an invocation exception. To get a meaningful message, we need the
-            // cause.
-            throw new AnalysisException(e.getCause.getMessage)
+        try {
+          f.newInstance(expressions : _*).asInstanceOf[Expression]
+        } catch {
+          // the exception is an invocation exception. To get a meaningful message, we need the
+          // cause.
+          case e: Exception => throw new AnalysisException(e.getCause.getMessage)
         }
       }
     }
 
     (name, (expressionInfo[T](name), builder))
   }
 
+  private def expressionWithAlias[T <: Expression](name: String)
+      (implicit tag: ClassTag[T]): (String, (ExpressionInfo, FunctionBuilder)) = {
+    val constructors = tag.runtimeClass.getConstructors
+      .filter(_.getParameterTypes.head == classOf[String])
+    assert(constructors.length == 1)
+    val builder = (expressions: Seq[Expression]) => {
+      val params = classOf[String] +: Seq.fill(expressions.size)(classOf[Expression])
+      val f = constructors.find(_.getParameterTypes.toSeq == params).getOrElse {
```

Comment:
here we not only check the # of arguments, but also argument types. Shall we only check the # of arguments and leave the type check to the `newInstance` call?

Reply:
Here we are checking only # of arguments. Because till here, every argument is of type Expression and also every function's constructor takes Expression(s) as argument. Except `BoolAnd` and `BoolOr`, for which `classOf[String]` is prepended to Seq.

`expression` is also checking only # of arguments. Validating arguments types is done by `checkInputDataTypes()`.

Reply:
> also every function's constructor takes Expression(s) as argument

We don't guarantee it. And we should fail with meaningful error message if it happens, instead of "Invalid number of arguments"

Reply:
Okay. I will update.